demos/ota/ota_demo_core_http/ota_demo_core_http.c

```diff
@@ -0,0 +1,2486 @@
+    ( void ) memset( &xResponse, 0, sizeof( xResponse ) );
+    ( void ) memset( &xRequestHeaders, 0, sizeof( xRequestHeaders ) );
+
+    /* Initialize the request object. */
+    xRequestInfo.pHost = pcServerHost;
+    xRequestInfo.hostLen = xServerHostLength;
+    xRequestInfo.pMethod = HTTP_METHOD_GET;
+    xRequestInfo.methodLen = sizeof( HTTP_METHOD_GET ) - 1;
+    xRequestInfo.pPath = pcPath;
+    xRequestInfo.pathLen = strlen( pcPath );
+
+    /* Set "Connection" HTTP header to "keep-alive" so that multiple requests
+     * can be sent over the same established TCP connection. */
+    xRequestInfo.reqFlags = HTTP_REQUEST_KEEP_ALIVE_FLAG;
+
+    /* Set the buffer used for storing request headers. */
+    xRequestHeaders.pBuffer = pucHttpUserBuffer;
+    xRequestHeaders.bufferLen = HTTP_USER_BUFFER_LENGTH;
+
+    xHttpStatus = HTTPClient_InitializeRequestHeaders( &xRequestHeaders,
+                                                       &xRequestInfo );
+
+    HTTPClient_AddRangeHeader( &xRequestHeaders, ulRangeStart, ulRangeEnd );
+
+    if( xHttpStatus == HTTPSuccess )
+    {
+        /* Initialize the response object. The same buffer used for storing
+         * request headers is reused here. */
+        xResponse.pBuffer = pucHttpUserBuffer;
+        xResponse.bufferLen = HTTP_USER_BUFFER_LENGTH;
+
+        /* Send the request and receive the response. */
+        xHttpStatus = HTTPClient_Send( &xTransportInterfaceHttp,
+                                       &xRequestHeaders,
+                                       NULL,
+                                       0,
+                                       &xResponse,
+                                       0 );
+    }
+    else
+    {
+        LogError( ( "Failed to initialize HTTP request headers: Error=%s.",
+                    HTTPClient_strerror( xHttpStatus ) ) );
+    }
+
+    if( xHttpStatus != HTTPSuccess )
+    {
+        if( ( xHttpStatus == HTTPNoResponse ) || ( xHttpStatus == HTTPNetworkError ) )
+        {
+            xReconnectRequired = true;
+        }
+        else
+        {
+            LogError( ( "HTTPClient_Send failed: Error=%s.",
+                        HTTPClient_strerror( xHttpStatus ) ) );
+
+            xReturnStatus = OtaHttpRequestFailed;
+        }
+    }
+    else
+    {
+        /* Check if reconnection required. */
+        if( xResponse.respFlags & HTTP_RESPONSE_CONNECTION_CLOSE_FLAG )
+        {
+            xReconnectRequired = true;
+        }
+
+        /* Handle the http response received. */
+        xReturnStatus = prvHandleHttpResponse( &xResponse );
+    }
+
+    if( xReconnectRequired == true )
+    {
+        /* End TLS session, then close TCP connection. */
+        ( void ) SecureSocketsTransport_Disconnect( &xNetworkContextHttp );
+
+        /* Try establishing connection to S3 server again. */
+        if( prvConnectToS3Server( &xNetworkContextHttp, NULL ) == EXIT_SUCCESS )
```

Comment:
Should this be trying to reconnect with exponential backoff instead of just once?